### ISSUE TITLE
Unable to configure Spawner.options_from_form, "missing 1 required positional argument"

### ISSUE THREAD

**Issue by rcthomas** [jupyterhub/jupyterhub#3789]
#### Expected behaviour

Back in #3225 `Spawner.options_from_form` was made configurable, so I expected to be able to do something like this in a config file:

```
async def options_from_form(spawner, form_data):
    ...
    return user_options

c.Spawner.options_from_form = options_from_form
```

#### Actual behaviour

When processing the form though, you get the following error:

```
[E 2022-02-09 02:28:34.484 JupyterHub pages:264] Failed to spawn single-user server with form
    Traceback (most recent call last):
      File "/usr/local/lib/python3.8/dist-packages/jupyterhub/handlers/pages.py", line 258, in _post
        options = await maybe_future(spawner.options_from_form(form_options))
    TypeError: options_from_form() missing 1 required positional argument: 'form_data'
```

It looks like I'm really supposed to declare `async def options_from_form(form_data)` in my configuration and not be able to pass the spawner?  I realize that the function isn't supposed to have side effects and maybe (?) the idea behind not letting a user pass `spawner` along was to discourage side-effects, but it's somewhat inconvenient.

Note that if you create a new Spawner class and customize `options_from_form` you get access to the spawner/self.  That discrepancy makes me think this is actually a bug.

### How to reproduce

This `jupyterhub_config.py` seems to reproduce the problem:

```
async def options_form(spawner):
    return "hello"

async def options_from_form(spawner, form_data):
    return {}

c.Spawner.options_form = options_form
c.Spawner.options_from_form = options_from_form
```

The reproducer was confirmed using the `jupyterhub/jupyterhub:2` Docker container.

I find that I'm able to get the behavior I expect by creating a `run_options_from_form` wrapper in the Spawner class following the style of the `pre_spawn_hook` etc. but this breaks the situation where the default `options_from_form` is used, or (I think) where `options_from_form` is defined as part of a `Spawner` class, not as a configurable.

**Comment by rcthomas** [comment]
This is kind of "neat" actually.

**Comment by rcthomas** [comment]
This approach works with the `load_config_file` line commented out or not, and is kind of more in line with how the other hooks work:

```
from traitlets import Callable, default
from traitlets.config import Application

def _default_thing(obj, x):
    return x

class A(Application):

    thing = Callable().tag(config=True)

    @default("thing")
    def _thing(self):
        return _default_thing

    def initialize(self, argv=None):
        self.load_config_file("config.py") # to test, comment/uncomment this line
        return

if __name__ == "__main__":
    A.launch_instance()
    a = A.instance()
    print(a.thing)
    print(a.thing(a, 2)) # Note that this works weird but this is how the other hooks (e.g. pre_spawn_hook) do it
```

I can make this a PR, changing how the default works in `spawner.py` and updating the call to `options_from_form` in the `pages.py` handler?

**Comment by rcthomas** [comment]
Closed by #3971
